Log input event details to verbose log

intgr opened the pull request:
I wanted to automate a few actions using 'adb shell input' commands, but
it requires pixel coordinates. Logging input event details from scrcpy is
helpful for achieving that.

intgr commented:
PS: Thanks a lot for developing and maintainig scrcpy. :cupid:

rom1v commented:
Thank you for your contribution.

I think that the log should not be in the "convert" method (wihch is a kind of util method), but in the caller.

Input events can be very verbose, they should be enabled in verbose mode only (`LOGV`) IMO.

For consistency, verbose logs could also be added for other input events (keyboard, touches). (This could be done separately later though.)

intgr commented:
:+1: I will try to make changes later today.

intgr commented:
Fixed your notes and added logging for finger events. Didn't implement keyboard event logging yet.

rom1v commented:
After reading your `LOGV` calls just before a `control_push_msg()`, I'm wondering if we shouldn't log all control messages (with [all its values](https://github.com/Genymobile/scrcpy/blob/d3d955f67b92d08bbee21bebce40fa072eea38f6/app/src/control_msg.h#L42-L74)).

For example, by adding a function `control_msg_log(const struct control_msg *msg)` called at the beginning of `control_push_msg()`. What do you think?

intgr commented:
Makes sense to me.

Do you think I should log `AMOTION_EVENT_ACTION_MOVE` events too or just up/down? For my use case, I would find logging all events a bit noisy.

For example, if I want to script an `adb shell input swipe` command, I need the start and end coordinates but not the intermediate ones. But I think it isn't a big deal.

Can you imagine any use cases, where logging intermediate events would be useful?

rom1v commented:
I would prefer to log all events sent to the device without distinction. If you want only some of them, it's easy to grep :)

intgr commented:
@rom1v I pushed a partial draft, if you happen to be around, it would be nice to hear if I'm going in the right direction with this.

I also noticed only now (embarrassing!) that scrcpy does not currently support a `SC_LOG_LEVEL_VERBOSE` log level, I suppose I need to add support for that as well?

rom1v commented on app/src/android/input.h:
```diff
@@ -91,6 +91,8 @@ enum android_input_event_type {
 /**
  * Key event actions.
+ *
+ * Also keep `android_keyevent_action_descriptions` updated.
```
This file is copied from AOSP: [`include/android/input.h`](https://android.googlesource.com/platform/frameworks/native/+/refs/heads/master/include/android/input.h). Please do not modify it.

rom1v commented on app/src/controller.c:
```diff
@@ -4,6 +4,34 @@
+#define ENUM_TO_DESCRIPTION(descriptions, value) \
+    ((size_t) (value) < (sizeof(descriptions)/sizeof(descriptions[0])) ? descriptions[value] : "???")
+
+static const char *android_keyevent_action_descriptions[] = {
```
```c
static const char *const android_keyevent_action_descriptions[] = { … }
```

rom1v commented on app/src/controller.c:
```diff
@@ -4,6 +4,34 @@
+ * out-of-bounds value.
+ */
+#define ENUM_TO_DESCRIPTION(descriptions, value) \
+    ((size_t) (value) < (sizeof(descriptions)/sizeof(descriptions[0])) ? descriptions[value] : "???")
```
There is a macro `ARRAY_SIZE()`.

rom1v commented on app/src/controller.c:
```diff
@@ -45,9 +73,67 @@ controller_destroy(struct controller *controller) {
+control_msg_log(const struct control_msg *msg) {
+    switch(msg->type) {
+        case CONTROL_MSG_TYPE_INJECT_KEYCODE:
+            LOGV("Event: key %s code=%u repeat=%u meta=%06x",
```
The actual type of an enum is implementation defined ([specs](https://port70.net/~nsz/c/c11/n1570.html#6.7.2.2p4)), so the value should be cast to a concrete integer type (`int` for example, with `"%d"` as format),

The `repeat` value is `uint32_t`, so the format must be `PRIu32`.

_Idem for other values._

```c
LOGV("Event: key %s code=%d repeat=%" PRIu32 " meta=%06lx",
    ENUM_TO_DESCRIPTION(android_keyevent_action_descriptions, msg->inject_keycode.action),
    (int) msg->inject_keycode.keycode,
    msg->inject_keycode.repeat,
    (long) msg->inject_keycode.metastate); // in theory, int may be as small as 16bits, so not sufficient
```
_(keep your own indentation, I just wrote it from the github interface.)_

rom1v reviewed (commented):
> I pushed a partial draft, if you happen to be around, it would be nice to hear if I'm going in the right direction with this.

Yes, looks good :+1: Thank you.

I added few remarks inline.

> I also noticed only now (embarrassing!) that scrcpy does not currently support a `SC_LOG_LEVEL_VERBOSE` log level

Oops.

> I suppose I need to add support for that as well?

Yes please, as a separate commit :)

intgr commented:
Thanks! My C has been getting a bit rusty. Pushed an updated version.

* I added `--verbosity=verbose` option in a separate commit
* I went through all format string arguments and should have correct types now.
* Moved logging code to `control_msg.c`
* Added remaining event types

I was trying to avoid building the Java part. Took me a while to realize that I need to downgrade, seems like it doesn't work with JDK 16 (got some weird error "module java.base does not "opens java.io" to unnamed module").

intgr commented on app/src/control_msg.c:
```diff
@@ -93,6 +128,74 @@ control_msg_serialize(const struct control_msg *msg, unsigned char *buf) {
+            break;
+        case CONTROL_MSG_TYPE_INJECT_TOUCH_EVENT: {
+            size_t action = msg->inject_touch_event.action & AMOTION_EVENT_ACTION_MASK;
+            LOGV("Input: touch %-4s position=%" PRIi32 "x%" PRIi32 " pressure=%g buttons=%06lx",
```
I omitted `pointer_id` from this message because it didn't didn't seem very useful.

rom1v replied:
I think you should log it, it allows to distinguish several simultaneous fingers on touch events.

intgr commented on server/src/main/java/com/genymobile/scrcpy/Server.java:
```diff
@@ -251,6 +251,7 @@ public void uncaughtException(Thread t, Throwable e) {
         Options options = createOptions(args);
 
         Ln.initLogLevel(options.getLogLevel());
+        Ln.v("Using verbose log level");
```
Should I drop this log and leave the `v()` method unused?

rom1v replied:
I think you should drop it.

intgr commented on app/src/control_msg.c:
```diff
@@ -93,6 +128,74 @@ control_msg_serialize(const struct control_msg *msg, unsigned char *buf) {
+            break;
+        case CONTROL_MSG_TYPE_INJECT_TOUCH_EVENT: {
+            size_t action = msg->inject_touch_event.action & AMOTION_EVENT_ACTION_MASK;
+            LOGV("Input: touch %-4s position=%" PRIi32 "x%" PRIi32 " pressure=%g buttons=%06lx",
```
The `%-4s` makes output more pleasantly aligned
```
VERBOSE: Input: touch down position=286x786 pressure=1 buttons=000001
VERBOSE: Input: touch move position=287x786 pressure=1 buttons=000001
VERBOSE: Input: touch move position=288x786 pressure=1 buttons=000001
VERBOSE: Input: touch up   position=288x786 pressure=0 buttons=000001
```

intgr commented on app/src/control_msg.c:
```diff
@@ -8,6 +8,41 @@
+ * Map an enum value to a string based on an array, without crashing on an
+ * out-of-bounds value.
+ */
+#define ENUM_TO_LABEL(labels, value) \
```
Renamed "description" -> "label" everywhere.

rom1v replied:
:+1: (or even `ENUM_TO_STRING`?)

intgr replied:
Yeah I slightly prefer "label" too :)

rom1v commented on app/src/cli.c:
```diff
@@ -184,7 +184,7 @@ scrcpy_print_usage(const char *arg0) {
 #endif
         "    -V, --verbosity value\n"
-        "        Set the log level (debug, info, warn or error).\n"
+        "        Set the log level (verbose, debug, info, warn or error).\n"
```
:+1:

Also in the manpage:

```diff
diff --git a/app/scrcpy.1 b/app/scrcpy.1
index 7b6f1faa..62dc9677 100644
--- a/app/scrcpy.1
+++ b/app/scrcpy.1
@@ -193,7 +193,7 @@ It requires to lock the video orientation (see --lock-video-orientation).
 
 .TP
 .BI "\-V, \-\-verbosity " value
-Set the log level ("debug", "info", "warn" or "error").
+Set the log level ("verbose", "debug", "info", "warn" or "error").
 
 Default is "info" for release builds, "debug" for debug builds.
 
```

rom1v commented:
Hi,

I allowed myself to work on this based on your branch. I did the following changes:
 - introduce a few more macros to simplify
 - log pointer id (either `mouse`, `vfinger` or a numeric id)

The logs highlighted that motion events were sent even if left-click was not pressed (but for example right-click or 4th button), so I fixed it: 2b7f19d2699ed1a16342605a72740a3fdc377460 :+1:

Here is a branch: [`logv`](https://github.com/Genymobile/scrcpy/commits/logv). Please review/test :wink:

I plan to do few more things before merging:
 - move all log level stuff to log.h/log.c
 - expose a function to know the current log level
 - do not call `control_msg_log()` if the current log level is not verbose.

rom1v commented:
> I plan to do few more things before merging:
> - move all log level stuff to log.h/log.c
> - expose a function to know the current log level
> - do not call `control_msg_log()` if the current log level is not verbose.

Done. I updated the `logv` branch.

intgr commented:
> I discovered, thanks to the logs, that motion events were sent even if left-click was not pressed

Thanks, cool to hear that there are other use cases/benefits of this work :)

intgr commented:
How do we proceed? Should I pull your changes to my branch? I'll do that for now.

rom1v commented:
> How do we proceed? Should I pull your changes to my branch?

Not necessarily, just tell me if it's ok or if some thkngs should be changed :) I'll merge manually.

intgr commented on app/src/controller.c:
```diff
@@ -48,6 +48,10 @@ controller_destroy(struct controller *controller) {
 bool
 controller_push_msg(struct controller *controller,
                       const struct control_msg *msg) {
+    if (sc_get_log_level() == SC_LOG_LEVEL_VERBOSE) {
```
I'd use `(sc_get_log_level() <= SC_LOG_LEVEL_VERBOSE)` even if we don't expect any lower log levels.

rom1v replied:
:+1:

intgr commented:
Other than that minor comment, this looks good to me.

> Not necessarily, just tell me if it's ok or if some thkngs should be changed :) I'll merge manually.

OK, I already pushed it, I'm sure you can sort it out :)

intgr commented on app/src/control_msg.c:
```diff
@@ -93,6 +140,94 @@ control_msg_serialize(const struct control_msg *msg, unsigned char *buf) {
+            if (id == POINTER_ID_MOUSE || id == POINTER_ID_VIRTUAL_FINGER) {
+                // string pointer id
+                LOG_CMSG("touch [id=%s] %-4s position=%" PRIi32 ",%" PRIi32
+                             " pressure=%f buttons=%06lx",
```
I used `%g` for pressure so it hides the fractional part when there are no digits. I guess you changed this intentionally?

`pressure=0.000000` vs `pressure=0`

rom1v replied:
Oh, yes, I forgot to discuss this change, the reason was that `%g` prints exponential notation when necessary, and I think we don't want that here:

```c
printf("%f %g\n", 0.0000001f, 0.0000001f);
```

prints:

```
0.000000 1e-07
```

But yes, `%g` is probably better in practice. I will change back to `%g`.

I also changed `position=500x400` to `position=500,400`, since `500x400` suggests a dimensions rather than a position.

rom1v replied:
Pushed to `logv`.

intgr replied:
Agreed, the `,` separator makes sense.

rom1v commented:
> OK, I already pushed it, I'm sure you can sort it out :)

OK, it's indeed better that you pushed it, it's more convenient for discussions :wink:

intgr commented:
Pushed again. Seems fine to me now :+1:

rom1v commented:
> Seems fine to me now +1

Thanks :+1:

Then let's merge into `dev` :tada:
